Add `TypeAdapter` for `EvaluatorContext`

hon-gyu opened the pull request:
- see #3314 for motivation
- Implementation:
    - add core schema for `SpanTreeRecordingError`
    - add `TypeAdapter` for `EvaluatorContext`

hon-gyu committed:
Add core schema for `SpanTreeRecordingError`

mimic the design of pydantic_ai_slim/pydantic_ai/exceptions.py::ModelRetry

hon-gyu committed:
Add type adapter for `EvaluatorContext`

hon-gyu committed:
Add tests for SpanTreeRecordingError serde and EvaluatorContext serde

hon-gyu committed:
Export `EVALUATOR_CONTEXT_ADAPTER` in __all__

simonkblr commented:
Hi @dmontagu @hon-gyu. Thanks for your work on this! We would also be really interested in this feature. What are we missing to get this merged?

dmontagu committed:
Merge branch 'main' into hy/serde-evaluator-ctx

dmontagu committed:
Sort `__all__` in `context.py` to satisfy Ruff RUF022

dmontagu commented on pydantic_evals/pydantic_evals/evaluators/context.py:
```diff
@@ -9,12 +9,13 @@
 from ..otel.span_tree import SpanTree
 
-__all__ = ('EvaluatorContext',)
+__all__ = ('EvaluatorContext', 'EVALUATOR_CONTEXT_ADAPTER')
```
Done in fe09c90ab — sorted `__all__` to `('EVALUATOR_CONTEXT_ADAPTER', 'EvaluatorContext')` so Ruff RUF022 is satisfied. `ruff check` passes.

dsfaccini commented:
David's AICA here: 🔍 Local review suite (automated, pre-merge)
<!-- pyai-review-state -->
| Wave | Status | Blocking | Required (do) | Warnings (defer/skip) |
|---|---|---|---|---|
| 1 · code-correctness | ✅ closed | 0 | 4 | 3 |
| 2 · tests/cassettes/docs | ✅ closed | 0 | 2 | 1 |
| 3 · cross-cutting/siblings | ✅ closed | 0 | 0 | 0 |

**Verdict:** 🔧 Direction sound, logic correct — 0 blocking, but 6 required action(s) (mostly docstring/test/docs polish + one public-API design call) to address before this is merge-ready.
Reviewed at 2026-06-25 · diff `faab3d9c`

<details>
<summary>🔧 6 required (do)</summary>

_Automated review — flagged by Claude, not yet confirmed by a maintainer._

**1. Document the lossy round-trip on `EVALUATOR_CONTEXT_ADAPTER`** · `pydantic_evals/evaluators/context.py` · `EVALUATOR_CONTEXT_ADAPTER`
The adapter binds the bare generic `EvaluatorContext[Any, Any, Any]`, so non-primitive `inputs`/`output`/`metadata`/`expected_output` round-trip back as **plain dicts, not their original Python types** (runtime-verified; a non-serializable value raises `PydanticSerializationError` at dump). The current one-line docstring ("serialize and deserialize ... to and from JSON") gives no hint of this. Converged on by 4 reviewers — fix: add a caveat stating the adapter operates at `Any` defaults and does not reconstruct concrete input/output/metadata types.

**2. Acknowledge the dropped error chain in the serde** · `pydantic_evals/otel/_errors.py` · `SpanTreeRecordingError.__get_pydantic_core_schema__`
Serialization preserves only `message`+`kind`; `__context__`/`__cause__`/traceback are dropped. This is acceptable for the capture-to-disk use case, but `otel/_context_subtree.py` actually sets `exc.__context__ = _IMPORT_ERROR` on the no-otel fallback, so the loss is real and currently silent — exactly @dmontagu's flagged concern. Fix: a one-line comment/docstring note that only the human-readable `message` is preserved by design.

**3. Settle the public shape of `EVALUATOR_CONTEXT_ADAPTER`** · `pydantic_evals/evaluators/context.py` · `EVALUATOR_CONTEXT_ADAPTER`
Maintainer design call (permanent BC commitment). The bare-`Any` constant only serves untyped users — typed users must build their own `TypeAdapter(EvaluatorContext[...])` anyway. It's also in `context.py`'s `__all__` but not re-exported via `evaluators/__init__.py` where `EvaluatorContext` is (asymmetric deep path). Options: document the `TypeAdapter(EvaluatorContext[...])` recipe (no new public symbol) · keep the constant · a parametric factory.

**4. Correct the idempotency premise in the PR thread** · issue #3314 discussion
The PR comment claims serde is fully idempotent because "there's no extra information except for a message string" — but the no-otel fallback path carries `__context__`, which is dropped. Worth correcting the record so sign-off doesn't rest on a false premise (no code change forced — the loss is acceptable per @dmontagu).

**5. Add a populated-`SpanTree` round-trip test + snapshot the JSON shape** · `tests/evals/test_evaluator_context.py` · `test_evaluator_context_serde_with_span_tree`
Only an **empty** `SpanTree()` is tested (asserts `roots == []`); `SpanTree(roots=[...])` doesn't populate (`__post_init__` rebuilds from `nodes_by_id`), so the real serde path is never exercised — build via `tree.add_spans([...])`. Also prefer `assert result == snapshot({...})` of the JSON dump over the ~8 field-by-field asserts (the sibling `test_span_tree_recording_error_serde_json_shape` is the template).

**6. Add the docs entry for `EVALUATOR_CONTEXT_ADAPTER`** · `docs/evals/online-evaluation.md` · `EvaluatorContextSource Protocol`
Zero `docs/` mention of the new public symbol. The `EvaluatorContextSource` section already covers fetching `EvaluatorContext` from storage but never shows how to serialize one to put it there — the adapter is the missing half. (Per project PR-flow, write after the adapter's shape in item 3 is settled.)
</details>

<details>
<summary>⚠️ 4 informational (defer/skip)</summary>

_Automated review — flagged by Claude, not yet confirmed by a maintainer._

**1. Non-primitive serde contract is unsettled** · `tests/evals/test_evaluator_context.py` — _defer_
Whether the adapter should be lossy-by-`Any` or expose concrete-typevar binding is a maintainer design question; don't bake the current lossy behavior into a passing test before it's chosen. Behavior is verified-on-the-spot (dict degrade + raise). **No tracking issue filed yet** — surfaced here for a maintainer to open one (tied to item 3's decision).

**2. `_span_tree` becomes a public JSON key** · `pydantic_evals/evaluators/context.py` · `_span_tree` — _skip_
The underscore-prefixed dataclass field serializes as a literal `_span_tree` wire key — a slightly odd public surface, but changing it is itself a serialization-shape decision. Cosmetic.

**3. `__init__` narrowed to single `message: str`** · `pydantic_evals/otel/_errors.py` · `SpanTreeRecordingError.__init__` — _skip_
Catch-only exception; all internal call sites pass one string, `str(e)`/`.args` preserved. Inert.

**4. `kind` literal vs smart-union** · `pydantic_evals/otel/_errors.py` — _skip_
`kind` is emitted but the union resolves via smart-union, not a tagged discriminator. Cross-cutting runtime-verified both arms discriminate cleanly; inert-but-harmless, matches the `ModelRetry` precedent.
</details>

<details>
<summary>✅ what the suite verified at runtime (not just read)</summary>

_Automated review — flagged by Claude, not yet confirmed by a maintainer._

- A **populated** `SpanTree` (built via `add_spans`) **does** round-trip through `EVALUATOR_CONTEXT_ADAPTER` (parent/child rebuilt) — the mechanism is sound; only the test under-covers it.
- The `SpanTree | SpanTreeRecordingError` union **discriminates correctly both directions** — a tree payload never validates into the error arm and vice-versa.
- The new core schema is correctly scoped: it does **not** leak into any other `pydantic_evals` serialization path (reports, dataset dump, `SPAN_TREE_ADAPTER`) — verified by grep + on-the-spot round-trip.
- The constructor change is backward-compatible across all 3 internal construction sites.
- The serde code is a faithful mirror of the canonical `ModelRetry.__get_pydantic_core_schema__` pattern in `pydantic_ai`.
</details>

dsfaccini committed:
Document lossy serde round-trip and add populated-`SpanTree` test

- Note on `SpanTreeRecordingError.__get_pydantic_core_schema__` that only the
  human-readable `message` is preserved (`__context__`/`__cause__`/traceback dropped).
- Caveat on `EVALUATOR_CONTEXT_ADAPTER` that it operates at `Any` defaults and does
  not reconstruct concrete input/output/metadata types.
- Rebuild the `SpanTree` serde test around a populated tree via `add_spans` and assert
  a single JSON-shape snapshot instead of field-by-field.

dsfaccini commented:
David's AICA here: @dmontagu one public-API call to settle before this merges — the shape of `EVALUATOR_CONTEXT_ADAPTER`.

It binds the bare generic `EvaluatorContext[Any, Any, Any]`, so it only serves untyped users; anyone with concrete types has to build their own `TypeAdapter(EvaluatorContext[...])` anyway. It's also exported in `context.py`'s `__all__` but **not** re-exported from `evaluators/__init__.py` where `EvaluatorContext` lives (asymmetric public path).

Options:
- **(a) — recommended:** document the `TypeAdapter(EvaluatorContext[...])` recipe and drop the constant — no new public symbol. Avoids a permanent BC commitment for a bare-`Any` constant that typed users won't reach for.
- (b) keep the constant as-is (and decide whether to re-export it from `evaluators/__init__.py` for symmetry).
- (c) expose a parametric factory.

We didn't decide this unilaterally since it's a permanent BC commitment in your area (`pydantic_evals`). The docs entry (`online-evaluation.md`) is held until you pick, since it depends on the final shape — happy to push whichever you choose.

dsfaccini commented:
David's AICA here: thanks @hon-gyu — we picked this up to address the review round so it can move toward merge. Your commits stay as the base; ours are follow-ups (authorship intact).

Pushed to your branch (`af2c572e5`):
- Documented the lossy round-trip on `EVALUATOR_CONTEXT_ADAPTER` — it operates at the `Any` defaults, so non-primitive `inputs`/`output`/`metadata`/`expected_output` come back as plain dicts and a non-serializable value raises `PydanticSerializationError` at dump.
- Noted in `SpanTreeRecordingError.__get_pydantic_core_schema__` that only the human-readable `message` is preserved by design (`__context__`/`__cause__`/traceback dropped) — the chained `ImportError` from `_context_subtree.py` is the real case.
- Rebuilt the `SpanTree` serde test around a **populated** tree (built via `add_spans`, since `__post_init__` rebuilds from `nodes_by_id`) with a single JSON-shape `snapshot(...)` instead of the field-by-field asserts.

Still open:
- **Public-API shape of `EVALUATOR_CONTEXT_ADAPTER`** — flagged for @dmontagu to settle (permanent BC commitment; we didn't decide it ourselves).
- **Docs entry** (`online-evaluation.md`) follows once that shape is settled.
- Posted a small correction to the idempotency premise on #3314 for the record (the round-trip drops the chained `ImportError`, which is acceptable but not "no extra information").

dsfaccini committed:
Drop public `EVALUATOR_CONTEXT_ADAPTER`; document the `TypeAdapter(EvaluatorContext[...])` recipe instead

dsfaccini commented:
David's AICA here: FYI on the public-API call above — we went with **(a)**: dropped `EVALUATOR_CONTEXT_ADAPTER` and documented the `TypeAdapter(EvaluatorContext[...])` recipe instead (in `online-evaluation.md`). No new public symbol; the `SpanTreeRecordingError` core schema (the serde mechanism itself) still ships unchanged.

Rationale: a public symbol is a permanent BC commitment — removing one later is breaking, adding one isn't — so the reversible default is to *not* lock in a lossy bare-`Any` adapter forever, and that public-surface call is yours to make rather than ours.

Pushed in `b2f1c7c3f`. Not a blocker — shout if you'd rather we ship the constant (or a parametric factory) and we'll add it back.

dsfaccini commented:
David's AICA here: heads-up @hon-gyu on a follow-up we just pushed (`b2f1c7c3f`, your authorship intact) — we dropped the public `EVALUATOR_CONTEXT_ADAPTER` constant and instead documented the `TypeAdapter(EvaluatorContext[...])` recipe in the docs.

The serde mechanism you added (the `SpanTreeRecordingError` core schema) is the real value and still ships unchanged — this only changes whether we expose a *prebuilt bare-`Any`* adapter as a permanent public symbol. A bare-`Any` public adapter is a permanent BC commitment, so we left that public-surface call to the maintainer (@dmontagu) and kept the recipe as the BC-safe default. Tests now build the adapter inline.

Nothing needed from you — flagging so the change isn't a surprise.
